### ISSUE TITLE
Incorrect type inference when a lambda is conditionally reassigned

### ISSUE THREAD

**Issue by wch** [microsoft/pyright#8229]
With pyright 1.1.369, I'm running into a problem where if a lambda function is reassigned within an `if` statement, it can cause an incorrect inferred type of an `async` function that calls that lambda. Best to just illustrate with an example:

```py
import asyncio


def foo(replace_inner: bool) -> None:
    inner = lambda: None

    async def wrapper():
        inner()

    wrapped_fn = wrapper()
    asyncio.create_task(wrapped_fn)

    if replace_inner:
        inner = lambda: None
```

When I run `pyright` on it, I get the following:

```
❯ pyright test.py
/Users/winston/test.py
  /Users/winston/test.py:11:25 - error: Argument of type "None" cannot be assigned to parameter "coro" of type "_CoroutineLike[_T@create_task]" in function "create_task"
    Type "None" is incompatible with type "_CoroutineLike[_T@create_task]"
      "None" is incompatible with "Generator[Any, None, _T@create_task]"
      "None" is incompatible with "Coroutine[Any, Any, _T@create_task]" (reportArgumentType)
1 error, 0 warnings, 0 informations 
```

When I hover over `warpped_fn`, it says the the type is `None`, when it should be `Coroutine[Any, Any, None]`:

<img width="384" alt="image" src="https://github.com/microsoft/pyright/assets/86978/72225918-f962-4da2-9775-11de4032cdd8">

Note that the following three modified versions of the code do not have this error (the change in each function has a comment above it):

```py
def foo2(replace_inner: bool) -> None:
    inner = lambda: None

    # Unlike in foo(), this inner function now has an explicit return type.
    async def wrapper() -> None:
        inner()

    wrapped_fn = wrapper()
    asyncio.create_task(wrapped_fn)

    if replace_inner:
        inner = lambda: None



def foo3(replace_inner: bool) -> None:
    inner = lambda: None

    async def wrapper():
        inner()

    wrapped_fn = wrapper()
    asyncio.create_task(wrapped_fn)

    # Unlike in foo(), the following line is not in an if statement.
    inner = lambda: None


def foo4(replace_inner: bool) -> None:
    # This time inner is a def function, instead of a lambda.
    def inner() -> None:
        return

    async def wrapper():
        inner()

    wrapped_fn = wrapper()
    asyncio.create_task(wrapped_fn)

    if replace_inner:
        inner = lambda: None
```

**Comment by erictraut** [comment]
Thanks for the bug report. I'm able to repro, and I've confirmed that it doesn't occur in 1.1.368. I'll investigate further.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.370
